src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,300 @@
+            R = rots[k, :, :]
+            if flag == 1:
+                R = J @ R @ J
+            regrot[k, :, :] = Q_mat.T @ R
+        aligned_rots = Rotation(K, matrices=regrot)
+        return aligned_rots
+
+    def register(self, rots_ref):
+        """
+         Estimate global orientation and return an aligned Rotation object.
+
+        :param rots_ref: The reference Rotation object to which we would like
+            to align with data matrices in the form of a n-by-3-by-3 array.
+        :return: an aligned Rotation object
+        """
+        Q_mat, flag = self.find_registration(rots_ref)
+        return self.apply_registration(Q_mat, flag)
+
+    def mse(self, rots_ref):
+        """
+        Calculate MSE between the estimated orientations to reference ones.
+
+        :param rots_reg: The estimated Rotation object after alignment
+             with data matrices in the form of a n-by-3-by-3 array.
+        :param rots_ref: The reference Rotation object.
+        :return: The MSE value between two sets of rotations.
+        """
+        aligned_rots = self.register(rots_ref)
+        rots_reg = aligned_rots.matrices
+        rots_ref = rots_ref.matrices
+        ensure(
+            rots_reg.shape == rots_ref.shape,
+            "Two sets of rotations must have same dimensions.",
+        )
+        K = rots_reg.shape[0]
+
+        diff = np.zeros(K)
+        mse = 0
+        for k in range(K):
+            diff[k] = norm(rots_reg[k, :, :] - rots_ref[k, :, :], ord="fro")
+            mse += diff[k] ** 2
+        mse = mse / K
+        return mse
+
+    def common_lines(self, i, j, ell):
+        """
+        Compute the common line induced by rotation matrices i and j.
+
+        :param i: The index of first rotation matrix of 3-by-3 array.
+        :param j: The index of second rotation matrix of 3-by-3 array.
+        :param ell: The total number of common lines.
+        :return: The common line indices for both first and second rotations.
+        """
+        r1 = self.matrices[i]
+        r2 = self.matrices[j]
+        ut = np.dot(r2, r1.T)
+        alpha_ij = np.arctan2(ut[2, 0], -ut[2, 1]) + np.pi
+        alpha_ji = np.arctan2(ut[0, 2], -ut[1, 2]) + np.pi
+
+        ell_ij = alpha_ij * ell / (2 * np.pi)
+        ell_ji = alpha_ji * ell / (2 * np.pi)
+
+        ell_ij = int(np.mod(np.round(ell_ij), ell))
+        ell_ji = int(np.mod(np.round(ell_ji), ell))
+
+        return ell_ij, ell_ji
+
+    @staticmethod
+    def uniform_random_angles(
+        n,
+        seed=None,
+        dtype=np.float32,
+    ):
+        """
+        Generate random 3D rotation angles in radians
+
+        :param n: The number of rotation angles to generate
+        :param seed: Random integer seed to use. If None,
+            the current random state is used.
+        :return: A n-by-3 ndarray of rotation angles in radians
+        """
+        # Generate random rotation angles, in radians
+        angles = np.zeros((n, 3), dtype=dtype)
```

Comment:
This line is not used. When you remove this, add `dtype` to the return. Add docstring for dtype.

Reply:
Done.